### ISSUE TITLE
Node GCM doesn't support Apple 'apns-collapse-id' or Android 'collapseKey' 

### ISSUE THREAD

**Issue by jbruggeman** [ToothlessGear/node-gcm#313]
Google has added 'apns-collapse-id' and 'collapseKey'. If you're using a non-legacy application it appears iOS 10+ devices will not get notifications. 

See https://firebase.google.com/docs/cloud-messaging/concept-options

**Comment by eladnava** [comment]
Hi @jbruggeman,
Thanks for the update!

The solution would be for `node-gcm` to support the new `apns-collapse-id` flag, right?

**Comment by jbruggeman** [comment]
@eladnava 

So, looking into it a bit further, google now has two versions of the API. node-gcm looks like it covers the legacy API, but not the new API, which hits a different endpoint. This new API was released within the last month.

Do you think it makes sense to update node-gcm to handle a separate endpoint or perhaps a new wrapper library should be created?

**Comment by eladnava** [comment]
Hi @jbruggeman,
That's interesting, can you please link to the article with the new API endpoint info?

It would probably make sense to update node-gcm as long as there are no breaking changes.

**Comment by jbruggeman** [comment]
@eladnava 

The new API is described here:
https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages

The old API is listed as 'Legacy HTTP Server Protocol'.

**Comment by eladnava** [comment]
Hi @jbruggeman,
Interesting, thanks for the links. 

Regarding collapse keys, `node-gcm` definitely supports those using the Legacy HTTP API parameter, see the Usage section in the README for syntax:
https://github.com/ToothlessGear/node-gcm#additional-message-options

I think that updating `node-gcm` to hit the new FCM HTTP v1 API should only be done once there is functionality which is not available in the Legacy HTTP Server Protocol.

@hypesystem fyi

**Comment by hypesystem** [comment]
I'm not quite sure I understand this correctly.

If it is true that there is no new functionality added (that there is no functionality current unsupported by node-gcm), I agree: we don't need to focus on a new version now.

If you are up for giving it a try, @jbruggeman, it might be cool to see how much would need to change in order to support the new API. I'm assuming that using the new API would still work for old applications, registering with the existing methods?

**Comment by jbruggeman** [comment]
@eladnava  Yes, it has a collapseKey, but it will not generate the APNS apns-collapse-id in the resulting message, which is required for collapsing on iOS 10+. Adding this key is only available using the FCM HTTP v1 API.

**Comment by eladnava** [comment]
@jbruggeman Cool, if there is indeed missing functionality we may be forced to update to the new API sooner than later.

Is there any information on the Internet that references the fact that the old Legacy HTTP API does not set the `apns-collapse-id` flag when the `collapse_key` parameter is specified?

**Comment by hypesystem** [comment]
A source for this would indeed be good, so we can document this current deficiency.

@jbruggeman do you have a subjective opinion on how much has changed between the API we use and the new one? I haven't had time to check it out, but it might be relevant to update what we do under the hood.
